docs/events/2023/docker-workshop.md

```diff
@@ -0,0 +1,235 @@
+
+### `CMD`
+
+The `CMD` command defines the command that is run when you run the docker image. This is how you define the command that runs your application.
+
+## Advanced `Dockerfile`s
+
+Because of the way that docker images are built, you can use them to build other docker images. This is useful for things like building a frontend and then building a backend that uses the frontend. You may build your application in one docker image then copy the build into another docker image and run it.
+
+This means that the programs that you used to build the application don't need to be included in the final version of your image. This is commonly called a multi-stage build and it is a great way to reduce the size of your docker images.
+
+```dockerfile
+# Stage 1 - Base (resources required by all layers)
+FROM node:20-alpine as base
+
+WORKDIR /app
+COPY package.json yarn.lock ./
+
+# Stage 2 - Build (dependencies required by the typescript compiler)
+FROM base as build
+
+RUN yarn install --frozen-lockfile
+COPY . .
+RUN yarn build
+RUN yarn install --production --frozen-lockfile
+
+# Stage 3 - Runner (runs the compiled code)
+FROM base as runner
+
+COPY --from=build /app/dist ./dist
+COPY --from=build /app/node_modules ./node_modules
+
+CMD ["node", "dist/app.js"]
+```
+
+This code is the same as the earlier example but uses a multi-stage build to bundle and transpile the code before running it. 
+
+This means that only the specific files copied from the build stage are included in the final image. This means that the final image is much smaller than it would be if you included the build tools in the final image.
+
+## `docker-compose`
+
+Docker compose is a tool that allows you to run multiple docker containers at the same time. This is useful for things like running a database and a web server that need to talk to one another at the same time.
+
+```yaml
+services:
+  tsnode:
+    image: ghcr.io/progsoc/docker-ts-2023:latest
+    ports:
+      - 3000:3000
+```
+
+This is an example of a docker compose file that runs the docker image that we built earlier. It exposes port 3000 on the host machine and runs the docker image.
+
+You might be able to see that the image url is similar to the way that `git` urls work. This is because docker images are stored in a similar way to git repositories. The `:latest` part tells docker to get the latest version of the image that has been pushed to the repository.
+
+You can change it to a specific version if you want to use a specific version of the image e.g. postgres 14 or 15.
+
+## `docker-compose` with multiple services
+
+```yaml
+services:
+  tsnode:
+    image: ghcr.io/progsoc/docker-ts-2023:latest
+    ports:
+      - 3000:3000
+  postgres:
+    image: postgres:14
+    ports:
+      - 5432:5432
+    environment:
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: postgres
+      POSTGRES_DB: postgres
+    volumes:
+      - postgres-data:/var/lib/postgresql/data
```

Comment:
I'd recommend avoiding teaching docker volumes, instead teaching volume mounts is much more practical. Just replace `postgres-data` with `./data` or something